cpp/velox/shuffle/VeloxSplitter.h

```diff
@@ -0,0 +1,370 @@
+
+  arrow::Status DoSplit(const facebook::velox::RowVector& rv);
+
+  std::string NextSpilledFileDir();
+
+  arrow::Result<std::shared_ptr<arrow::ipc::IpcPayload>> GetSchemaPayload();
+
+  uint32_t CalculatePartitionBufferSize(const facebook::velox::RowVector& rv);
+
+  arrow::Status AllocatePartitionBuffers(uint32_t partition_id, uint32_t new_size);
+
+  arrow::Status AllocateBufferFromPool(std::shared_ptr<arrow::Buffer>& buffer, uint32_t size);
+
+  arrow::Status AllocateNew(uint32_t partition_id, uint32_t new_size);
+
+  arrow::Status CreateRecordBatchFromBuffer(uint32_t partition_id, bool reset_buffers);
+
+  arrow::Status CacheRecordBatch(uint32_t partition_id, const arrow::RecordBatch& rb);
+
+  arrow::Status SplitFixedWidthValueBuffer(const facebook::velox::RowVector& rv);
+
+  arrow::Status SplitBoolType(const uint8_t* src_addr, const std::vector<uint8_t*>& dst_addrs);
+
+  arrow::Status SplitValidityBuffer(const facebook::velox::RowVector& rv);
+
+  arrow::Status SplitBinaryArray(const facebook::velox::RowVector& rv);
+
+  template <typename T>
+  arrow::Status SplitFixedType(const uint8_t* src_addr, const std::vector<uint8_t*>& dst_addrs) {
+    std::transform(
+        dst_addrs.begin(),
+        dst_addrs.end(),
+        partition_buffer_idx_base_.begin(),
+        partition_buffer_idx_offset_.begin(),
+        [](uint8_t* x, uint32_t y) { return x + y * sizeof(T); });
+
+    for (uint32_t pid = 0; pid < num_partitions_; ++pid) {
+      auto dst_pid_base = reinterpret_cast<T*>(partition_buffer_idx_offset_[pid]);
+      auto pos = partition_2_row_offset_[pid];
+      auto end = partition_2_row_offset_[pid + 1];
+      for (; pos < end; ++pos) {
+        auto row_id = row_offset_2_row_id_[pos];
+        *dst_pid_base++ = reinterpret_cast<const T*>(src_addr)[row_id]; // copy
+      }
+    }
+    return arrow::Status::OK();
+  }
+
+  arrow::Status SplitBinaryType(
+      uint32_t binary_idx,
+      const facebook::velox::FlatVector<facebook::velox::StringView>& src,
+      std::vector<BinaryBuff>& dst);
+
+  arrow::Status SplitListArray(const facebook::velox::RowVector& rv);
+
+  arrow::Result<int32_t> SpillLargestPartition(int64_t* size);
+
+  arrow::Status SpillPartition(uint32_t partition_id);
+
+ protected:
+  bool support_avx512_ = false;
+
+  uint32_t num_partitions_ = 0;
+
+  // options
+  SplitOptions options_;
+
+  // the first column may be stripped if hash split
+  std::shared_ptr<arrow::Schema> schema_;
+
+  // store arrow column types
+  std::vector<std::shared_ptr<arrow::DataType>> arrow_column_types_; // column_type_id_
+
+  // store velox column types
+  std::vector<std::shared_ptr<const facebook::velox::Type>> velox_column_types_;
+
+  // write options for tiny batches
+  arrow::ipc::IpcWriteOptions tiny_batch_write_options_;
+
+  // Row ID -> Partition ID
+  // subscript: Row ID
+  // value: Partition ID
+  // TODO: rethink, is uint16_t better?
+  std::vector<uint32_t> row_2_partition_; // note: partition_id_
```

Comment:
why naming these fields as `xxx_2_xxx` ? it is not a map, why not just `partitionIds` ?

Reply:
Feel free to update for more reasonable names.